rust/src/operations/update.rs

```diff
@@ -0,0 +1,921 @@
+//! Update records from a Delta Table for records statisfy a predicate
+//!
+//! When a predicate is not provided then all records are updated from the Delta
+//! Table. Otherwise a scan of the Delta table is performed to mark any files
+//! that contain records that satisfy the predicate. Once they are determined
+//! then column values are updated with new values provided by the user
+//!
+//!
+//! Predicates MUST be deterministic otherwise undefined behaviour may occur during the
+//! scanning and rewriting phase.
+//!
+//! # Example
+//! ```rust ignore
+//! let table = open_table("../path/to/table")?;
+//! let (table, metrics) = UpdateBuilder::new(table.object_store(), table.state)
+//!     .with_predicate(col("col1").eq(lit(1)))
+//!     .with_update("value", col("value") + lit(20))
+//!     .await?;
+//! ````
+
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+    time::{Instant, SystemTime, UNIX_EPOCH},
+};
+
+use arrow::datatypes::Schema as ArrowSchema;
+use arrow_array::RecordBatch;
+use arrow_cast::CastOptions;
+use arrow_schema::{Field, SchemaRef};
+use datafusion::{
+    execution::context::SessionState,
+    physical_plan::{
+        metrics::{ExecutionPlanMetricsSet, MetricBuilder, MetricsSet},
+        projection::ProjectionExec,
+        ExecutionPlan, RecordBatchStream, SendableRecordBatchStream,
+    },
+    prelude::SessionContext,
+};
+use datafusion_common::Result as DataFusionResult;
+use datafusion_common::{Column, DFSchema, ScalarValue};
+use datafusion_expr::{case, col, lit, when, Expr};
+use datafusion_physical_expr::{
+    create_physical_expr,
+    expressions::{self},
+    PhysicalExpr,
+};
+use futures::{future::BoxFuture, Stream, StreamExt};
+use parquet::file::properties::WriterProperties;
+use serde_json::{Map, Value};
+
+use crate::{
+    action::{Action, DeltaOperation, Remove},
+    delta_datafusion::{find_files, parquet_scan_from_actions, register_store},
+    storage::{DeltaObjectStore, ObjectStoreRef},
+    table_state::DeltaTableState,
+    DeltaResult, DeltaTable, DeltaTableError,
+};
+
+use super::{transaction::commit, write::write_execution_plan};
+
+/// Used to represent user input of either a Datafusion expression or string expression
+pub enum Expression {
+    /// Datafusion Expression
+    E(Expr),
+    /// String Expression
+    S(String),
+}
```

Comment:
nit: please don't use single character variable names

```suggestion
pub enum Expression {
    /// Datafusion Expression
    DataFusion(Expr),
    /// String Expression
    String(String),
}
```